### PR TITLE
Context Refactor 

### PR DESCRIPTION
**Pull request by vibhatha**

### CODE REVIEW

**Comment by supunkamburugamuve** [comment]
Two quick comments.

1. We should always use std::shared_ptr, std::make_shared etc
2 We need to change methods to create context to return a shared pointer, std::shared_ptr<Context> ctx = cylon::CylonContext::InitDistributed(mpi_config);

**Comment by vibhatha** [comment]
@supunkamburugamuve I modified the methods to return shared_ptr and also used make_shared in creation. 
Resolving some issues in Java bindings.

File: cpp/src/cylon/table.hpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ class Table {
   /**
    * Tables can only be created using the factory methods, so the constructor is private
    */
-  Table(std::shared_ptr<arrow::Table> &tab, cylon::CylonContext *ctx)
+  Table(std::shared_ptr<arrow::Table> &tab, const shared_ptr<cylon::CylonContext> &ctx)
```
Comment (nirandaperera) [inline cpp/src/cylon/table.hpp:44]: wouldn't `const` be a problem here?

File: cpp/src/cylon/table_api.cpp [hunk 2]
```diff
@@ -274,10 +274,10 @@ Status Subtract(CylonContext *ctx,
   return status;
 }
 
-Status Intersect(CylonContext *ctx,
-                 const std::string &table_left,
-                 const std::string &table_right,
-                 const std::string &dest_id) {
+Status Intersect(shared_ptr<CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:277]: looks like we'll have to fix the indentation in this file

File: cpp/src/cylon/table_api.cpp [hunk 3]
```diff
@@ -246,10 +246,10 @@ Status VerifyTableSchema(const std::shared_ptr<arrow::Table> &ltab,
 }
 
 
-Status Union(CylonContext *ctx,
-             const std::string &table_left,
-             const std::string &table_right,
-             const std::string &dest_id) {
+Status Union(shared_ptr<CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:249]: let's use `shared_ptr<CylonContext> &`

File: cpp/src/cylon/table_api.cpp [hunk 4]
```diff
@@ -260,10 +260,10 @@ Status Union(CylonContext *ctx,
   return status;
 }
 
-Status Subtract(CylonContext *ctx,
-                const std::string &table_left,
-                const std::string &table_right,
-                const std::string &dest_id) {
+Status Subtract(shared_ptr<CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:263]: let's use `shared_ptr<CylonContext> &`

File: cpp/src/cylon/table_api.cpp (hunk 2 above)
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:277]: let's use `shared_ptr<CylonContext> &`

File: cpp/src/cylon/table_api.cpp [hunk 5]
```diff
@@ -288,10 +288,10 @@ Status Intersect(CylonContext *ctx,
   return status;
 }
 
-Status DistributedUnion(CylonContext *ctx,
-                        const std::string &table_left,
-                        const std::string &table_right,
-                        const std::string &dest_id) {
+Status DistributedUnion(shared_ptr<CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:291]: let's use `shared_ptr<CylonContext> &`

File: cpp/src/cylon/table_api.cpp [hunk 6]
```diff
@@ -302,10 +302,10 @@ Status DistributedUnion(CylonContext *ctx,
   return status;
 }
 
-Status DistributedSubtract(CylonContext *ctx,
-                           const std::string &table_left,
-                           const std::string &table_right,
-                           const std::string &dest_id) {
+Status DistributedSubtract(shared_ptr<cylon::CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:305]: let's use `shared_ptr<CylonContext> &`

File: cpp/src/cylon/table_api.cpp [hunk 7]
```diff
@@ -316,10 +316,10 @@ Status DistributedSubtract(CylonContext *ctx,
   return status;
 }
 
-Status DistributedIntersect(CylonContext *ctx,
-                            const std::string &table_left,
-                            const std::string &table_right,
-                            const std::string &dest_id) {
+Status DistributedIntersect(shared_ptr<CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.cpp:319]: let's use `shared_ptr<CylonContext> &`

File: cpp/src/cylon/table_api.hpp [hunk 8]
```diff
@@ -33,67 +33,67 @@ namespace cylon {
...
 
 Status WriteCSV(const std::string &id, const std::string &path,
                 const cylon::io::config::CSVWriteOptions &options = cylon::io::config::CSVWriteOptions());
 
-Status JoinTables(CylonContext *ctx,
-                  const std::string &table_left,
-                  const std::string &table_right,
-                  cylon::join::config::JoinConfig join_config,
-                  const std::string &dest_id);
+Status JoinTables(shared_ptr<cylon::CylonContext> ctx,
```
Comment (nirandaperera) [inline cpp/src/cylon/table_api.hpp:49]: we'll have to 
1. fix indentation 
2. use `shared_ptr<CylonContext> & ctx`

File: cpp/src/cylon/ctx/cylon_context.cpp [hunk 9]
```diff
@@ -22,16 +22,16 @@
...
-  return new CylonContext(false);
+shared_ptr<CylonContext> CylonContext::Init() {
+  return make_shared<CylonContext>(false);
 }
 CylonContext::CylonContext(bool distributed) {
   this->is_distributed = distributed;
 }
 
-CylonContext *CylonContext::InitDistributed(net::CommConfig *config) {
+shared_ptr<CylonContext> CylonContext::InitDistributed(net::CommConfig *config) {
```
Comment (nirandaperera) [inline cpp/src/cylon/ctx/cylon_context.cpp:32]: I think we need to do the same `shared_ptr` change for `net::CommConfig` as well. @supunkamburugamuve WDYT?

File: cpp/src/cylon/ctx/cylon_context.cpp [hunk 10]
```diff
@@ -22,16 +22,16 @@
...
   if (config->Type() == net::CommType::MPI) {
-    auto ctx = new CylonContext(true);
+    auto ctx = make_shared<CylonContext>(true);
     ctx->communicator = new net::MPICommunicator();
```
Comment (nirandaperera) [inline cpp/src/cylon/ctx/cylon_context.cpp:35]: we'd need a shared pointer for the communicator as well.

File: cpp/src/examples/join_inmem_datagen_example.cpp [hunk 11]
```diff
@@ -24,13 +24,13 @@
 #include <mpi.h>
 
 void create_binary_table(char *const *argv,
-                         cylon::CylonContext *ctx,
+						 shared_ptr<cylon::CylonContext> &ctx,
```
Comment (nirandaperera) [inline cpp/src/examples/join_inmem_datagen_example.cpp:27]: let's fix indentations in this file

**Review by nirandaperera — COMMENTED** [review]
added some comments in line

File: cpp/src/cylon/ctx/cylon_context.cpp (hunk 9 above)
Comment (vibhatha) [reply cpp/src/cylon/ctx/cylon_context.cpp:32]: There are many places, which is not consistent. I think it is better to take a look at the whole source and decide them and refactor now. In this PR or another. I can help.

**Comment by supunkamburugamuve** [comment]
@vibhatha it seems the test cases need to be updated

**Comment by chathurawidanage** [comment]
This PR has been merged to below PR.

https://github.com/cylondata/cylon/pull/135

**Pull request closed without merging**